### PR TITLE
Update core functions to support multiple mime types

### PR DESCRIPTION
**Pull request by jjgrainger**
Trac ticket: https://core.trac.wordpress.org/ticket/56349

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/post.php [hunk 1]
```diff
@@ -716,13 +716,23 @@ function create_initial_post_types() {
...
  *
- * @param int  $attachment_id Attachment ID.
- * @param bool $unfiltered    Optional. Whether to apply filters. Default false.
+ * @param int    $attachment_id Attachment ID.
+ * @param bool   $unfiltered    Optional. Whether to apply filters. Default false.
+ * @param string $unfiltered    Optional. The file MIME type to return.
  * @return string|false The file path to where the attached file should be, false otherwise.
  */
-function get_attached_file( $attachment_id, $unfiltered = false ) {
+function get_attached_file( $attachment_id, $unfiltered = false, $mime_type = null ) {
```
Comment (felixarntz) [inline src/wp-includes/post.php:724]: After giving this another thought, I don't think we should add the `$mime_type` parameter to this function. The parameter's purpose right now really revolves around images, however this function is for _any_ attachment, not only images.

I think we should revert the changes in this function, also given that having the parameter here isn't crucial for anything else in this PR.

File: src/wp-includes/media.php [hunk 2]
```diff
@@ -240,7 +241,7 @@ function image_downsize( $id, $size = 'medium' ) {
 		$is_intermediate = true;
 	} elseif ( 'thumbnail' === $size && ! empty( $meta['thumb'] ) && is_string( $meta['thumb'] ) ) {
 		// Fall back to the old thumbnail.
-		$imagefile = get_attached_file( $id );
+		$imagefile = get_attached_file( $id, $mime_type );
```
Comment (felixarntz) [inline src/wp-includes/media.php:244]: Related to my other comment on `get_attached_file()`, this is the only usage of `get_attached_file()` that uses the new parameter. I think we can simply remove this change from here though since the actual file name for `$imagefile` isn't even used in this function. So it doesn't really matter in which MIME type the file name is requested.

File: src/wp-includes/media.php [hunk 3]
```diff
@@ -176,9 +176,10 @@ function image_hwstring( $width, $height ) {
  *
  * @since 2.5.0
```
Comment (felixarntz) [inline src/wp-includes/media.php:177]: Adding a new parameter should always be documented here, like
```suggestion
 * @since 2.5.0
 * @since 6.1.0 Added the $mime_type parameter.
```

Please also add such an additional `@since` annotation for every other function in this PR where you're adding the parameter.

File: src/wp-includes/media.php [hunk 4]
```diff
@@ -833,6 +835,14 @@ function image_get_intermediate_size( $post_id, $size = 'thumbnail' ) {
 		$data['url']  = path_join( dirname( $file_url ), $data['file'] );
 	}
 
+	// Use the correct MIME type if available.
+	if ( ! empty( $mime_type ) && isset( $data['sources'][ $mime_type ]['file'] ) && $mime_type !== $data['mime-type'] ) {
+		$data['file']      = str_replace( wp_basename( $data['file'] ), $data['sources'][ $mime_type ]['file'], $data['file'] );
+		$data['path']      = str_replace( wp_basename( $data['path'] ), $data['sources'][ $mime_type ]['file'], $data['path'] );
+		$data['url']       = str_replace( wp_basename( $data['url'] ), $data['sources'][ $mime_type ]['file'], $data['url'] );
+		$data['mime-type'] = $mime_type;
+	}
```
Comment (felixarntz) [inline src/wp-includes/media.php:844]: Let's in any case unset `$data['sources']` here since it's not part of the expected return type and also doesn't need to be.

Also I don't think there is a need to include `mime-type` in the data as you're doing above, since that's not part of the response either. _If_ we want to add it for some reason, it should be added in all circumstances and also added to the `@return` documentation for this function. But I don't think that's actually needed.

File: src/wp-includes/media.php [hunk 5]
```diff
@@ -939,7 +949,9 @@ function wp_get_registered_image_subsizes() {
  * @param int          $attachment_id Image attachment ID.
  * @param string|int[] $size          Optional. Image size. Accepts any registered image size name, or an array of
  *                                    width and height values in pixels (in that order). Default 'thumbnail'.
- * @param bool         $icon          Optional. Whether the image should fall back to a mime type icon. Default false.
+ * @param array|bool   $args          Optional. Additional arguments for the image to return. For backward
+ *                                    compatibility boolean can be used to determine whether the image should
+ *                                    fall back to a mime type icon. Default false.
```
Comment (felixarntz) [inline src/wp-includes/media.php:954]: Please review other WordPress core functions that accept `$args` for how they document all individual supported keys for the array. We need to follow that here and thus expand this parameter documentation.

File: src/wp-includes/media.php [hunk 6]
```diff
@@ -989,7 +996,7 @@ function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $icon
 	 *                                    an array of width and height values in pixels (in that order).
 	 * @param bool         $icon          Whether the image should be treated as an icon.
 	 */
-	return apply_filters( 'wp_get_attachment_image_src', $image, $attachment_id, $size, $icon );
+	return apply_filters( 'wp_get_attachment_image_src', $image, $attachment_id, $size, $args );
```
Comment (felixarntz) [inline src/wp-includes/media.php:999]: This is a breaking change and therefore must not be made. Let's keep passing the value of `$icon` here (a boolean). If `$args['icon']` is not set, that would simply be `false`.

File: src/wp-includes/media.php [hunk 7]
```diff
@@ -949,27 +961,22 @@ function wp_get_registered_image_subsizes() {
...
+		$mime_type = $args['mime_type'];
+	}
 
-				$src_file               = $icon_dir . '/' . wp_basename( $src );
-				list( $width, $height ) = wp_getimagesize( $src_file );
-			}
-		}
+	if ( is_array( $args ) && isset( $args['icon'] ) || is_bool( $args ) ) {
+		$icon = isset( $args['icon'] ) ? $args['icon'] : $args;
+	}
```
Comment (felixarntz) [inline src/wp-includes/media.php:974]: This is overly complicated as there is a duplicate `isset` check. Also, we should be more forgiving for the backward compatibility fix, any non-array `$args` can reliably be interpreted as `$icon`.
```suggestion
	if ( is_array( $args ) && isset( $args['icon'] ) ) {
		$icon = $args['icon'];
	} else if ( ! is_array( $args ) ) {
		$icon = (bool) $args;
	}
```

File: src/wp-includes/media.php [hunk 8]
```diff
@@ -1008,7 +1015,9 @@ function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $icon
  * @param int          $attachment_id Image attachment ID.
  * @param string|int[] $size          Optional. Image size. Accepts any registered image size name, or an array
  *                                    of width and height values in pixels (in that order). Default 'thumbnail'.
- * @param bool         $icon          Optional. Whether the image should be treated as an icon. Default false.
+ * @param array|bool   $args          Optional. Additional arguments for the image to return. For backward
+ *                                    compatibility boolean can be used to determine whether the image should
+ *                                    fall back to a mime type icon. Default false.
```
Comment (felixarntz) [inline src/wp-includes/media.php:1020]: See my above comment on documenting the possible keys more extensively, following core best practices.

File: src/wp-includes/media.php [hunk 9]
```diff
@@ -1128,12 +1139,14 @@ function wp_get_attachment_image( $attachment_id, $size = 'thumbnail', $icon = f
  * @param int          $attachment_id Image attachment ID.
  * @param string|int[] $size          Optional. Image size. Accepts any registered image size name, or an array of
  *                                    width and height values in pixels (in that order). Default 'thumbnail'.
- * @param bool         $icon          Optional. Whether the image should be treated as an icon. Default false.
+ * @param array|bool   $args          Optional. Additional arguments for the image to return. For backward
+ *                                    compatibility boolean can be used to determine whether the image should
+ *                                    fall back to a mime type icon. Default false.
```
Comment (felixarntz) [inline src/wp-includes/media.php:1144]: See above.

File: src/wp-includes/media.php [hunk 10]
```diff
@@ -5564,3 +5577,61 @@ function wp_increase_content_media_count( $amount = 1 ) {
 
 	return $content_media_count;
 }
+
+/**
+ * Retrieves an attachments preview image or icon source.
+ *
+ * @since n.e.x.t
```
Comment (felixarntz) [inline src/wp-includes/media.php:5584]: `n.e.x.t` does not work for WP core. Please use the actual targeted version (`6.1.0`). Also in other instances in this PR.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@jjgrainger The PR looks solid so far! I left a few comments, but nothing too major.

One high level feedback though that is crucial to address: Please also add test coverage. At a minimum we should have some coverage for `wp_get_attachment_preview_src()`, as well as at least one test for the new `$args` parameter where it was added.

**Comment by adamsilverstein** [comment]
@jjgrainger - Since we [reverted multi-mime support](https://core.trac.wordpress.org/changeset/54085) this is not currently needed, though it will come in handy if we reintroduce the feature so thanks for your work here! Closing for now.
